File: src/frontend/src/types/paymentSourceDocument.ts
import { OcrDocumentClassification } from '../lib/documentClassificationDecision';
import { SourceDocumentType } from '../lib/sourceDocumentType';

/**
 * A document that originates a PAYMENT request.
 *
 * <p>PAYMENT only. Quotation Management keeps one document per quotation and nothing here reaches
 * it.</p>
 */
export interface PaymentSourceDocumentDto {
    id: string;
    /** Backend-assigned. Never renumbered after a removal — "Documento 2" must keep its meaning. */
    sequenceNumber: number;

    attachmentId: string;
    attachmentFileName?: string | null;
    attachmentStorageReference?: string | null;

    supplierId?: number | null;
    supplierNameSnapshot?: string | null;
    supplierTaxIdSnapshot?: string | null;
    plantId?: number | null;
    plantName?: string | null;

    /** One of the seven approved values; see SOURCE_DOCUMENT_TYPES. */
    sourceDocumentType?: SourceDocumentType | string | null;
    documentNumber?: string | null;
    documentSeries?: string | null;
    documentDate?: string | null;
    dueDate?: string | null;
    currency?: string | null;

    netAmount?: number | null;
    taxAmount?: number | null;
    grossAmount?: number | null;
    /** Sum of this document's active items — the number the gross must agree with. */
    itemsTotal: number;

    ocrSuggestion?: string | null;
    ocrConfidence?: number | null;
    ocrTitleFound?: string | null;
    ocrEvidenceJson?: string | null;
    ocrConflictingEvidenceJson?: string | null;
    classificationSource?: string | null;
    classificationSuggestionSource?: 'OCR' | 'FALLBACK' | null;
    classificationConflictAcknowledged: boolean;
    classificationJustification?: string | null;

    classificationReviewedByFinance: boolean;
    classificationReviewedByUserId?: string | null;
    classificationReviewedByName?: string | null;
    classificationReviewedAtUtc?: string | null;

    requiresOperationInvoice: boolean;
    requiresAdvanceRegularization: boolean;
    requiresFinanceClassificationReview: boolean;

    isVoided: boolean;
    voidReason?: string | null;

    items: PaymentSourceDocumentItemDto[];

    /** Why this document is not yet submittable. Empty when it is. */
    validationMessages: string[];
    isValid: boolean;

    /** Echoed on update — a stale edit is reported, never merged. */
    rowVersion?: string | null;
}

export interface PaymentSourceDocumentItemDto {
    id: string;
    lineNumber: number;
    description: string;
    quantity: number;
    unitId?: number | null;
    unitCode?: string | null;
    unitPrice: number;
    discountAmount?: number | null;
    ivaRateId?: number | null;
    totalAmount: number;
    plantId?: number | null;
    supplierId?: number | null;
    requestPoGroupId?: string | null;
}

/** Everything the payment screen needs about a request's origin documents. */
export interface PaymentSourceDocumentsSummaryDto {
    requestId: string;
    documents: PaymentSourceDocumentDto[];
    /** Voided documents, kept visible because their decisions were audited. */
    voidedDocuments: PaymentSourceDocumentDto[];
    requestTotal: number;
    currency?: string | null;

    /**
     * This request uses the multi-source-document model. TRUE with zero documents means a NEW draft
     * awaiting its first one — the collection must render. FALSE with zero documents means a
     * historical request that keeps the legacy single-document layout. A row count alone cannot
     * tell those apart, which is why this is persisted server-side rather than inferred.
     */
    usesMultiDocumentModel: boolean;

    canEditDocuments: boolean;
    editBlockedReason?: string | null;

    /** Request-level problems: no documents at all, mixed currencies. */
    requestValidationMessages: string[];

    /**
     * Non-blocking: the same supplier appears with different document types. Submission is never
     * refused because of it, and no justification is owed.
     */
    mixedTypeNotice?: string | null;

    canSubmit: boolean;
}

/** Create or replace a source document. */
export interface SavePaymentSourceDocumentDto {
    attachmentId?: string | null;

    supplierId?: number | null;
    supplierTaxIdSnapshot?: string | null;
    plantId?: number | null;

    sourceDocumentType?: string | null;
    documentNumber?: string | null;
    documentSeries?: string | null;
    documentDate?: string | null;
    dueDate?: string | null;
    currency?: string | null;

    netAmount?: number | null;
    taxAmount?: number | null;
    grossAmount?: number | null;

    ocrSuggestion?: string | null;
    ocrConfidence?: number | null;
    ocrTitleFound?: string | null;
    ocrEvidenceJson?: string | null;
    ocrConflictingEvidenceJson?: string | null;
    classificationSource?: string | null;
    classificationSuggestionSource?: 'OCR' | 'FALLBACK' | null;
    classificationConflictAcknowledged?: boolean | null;
    classificationJustification?: string | null;

    /** Required on update. Absent or stale is reported as a conflict, never merged. */
    rowVersion?: string | null;
}

export interface VoidPaymentSourceDocumentDto {
    reason?: string | null;
    rowVersion?: string | null;
}

/**
 * A 409 from an update or void. Carries the server's current view so the UI can show what moved
 * instead of leaving the user to guess.
 */
export interface PaymentSourceDocumentConflictDto {
    title: string;
    detail: string;
    status: number;
    current?: PaymentSourceDocumentDto | null;
}

/**
 * One document's OCR reading, held per document.
 *
 * <p>The single most important property of this feature: <b>nothing about a document's reading may
 * live in a request-level variable.</b> Uploading Documento 2 must not touch Documento 1.</p>
 */
export interface PaymentDocumentOcrState {
    isProcessing: boolean;
    classification: OcrDocumentClassification | null;
    /** Persistent, not a toast — a failed read must stay on screen until it is dealt with. */
    error: string | null;
}


/**
 * The answer to "may this file be attached?", asked before uploading or reading it.
 *
 * <p><c>SAME_REQUEST</c> is always <c>BLOCK</c>: there is no reading of a file already on this
 * request that would make paying the same invoice twice correct, so there is nothing to acknowledge.
 * <c>OTHER_REQUEST</c> is <c>WARN</c> — legitimate cases exist and the user decides.</p>
 */
export interface SourceDocumentDuplicateResult {
    isDuplicate: boolean;
    duplicateScope: 'NONE' | 'SAME_REQUEST' | 'OTHER_REQUEST';
    outcome: 'NONE' | 'BLOCK' | 'WARN';

    conflictingDocumentId?: string | null;
    conflictingSequenceNumber?: number | null;
    conflictingDocumentNumber?: string | null;
    supplierName?: string | null;

    /** Present only when the user may see the other request. */
    requestNumber?: string | null;
    requestId?: string | null;
    uploadedBy?: string | null;
    createdAtUtc?: string | null;
}

File: src/backend/AlplaPortal.Domain/Entities/PaymentSourceDocument.cs
using System;
using System.Collections.Generic;

namespace AlplaPortal.Domain.Entities;

/// <summary>
/// One document that originates a PAYMENT request.
///
/// <para><b>PAYMENT only.</b> Quotation Management keeps one document per quotation, carried on
/// <see cref="Quotation.DocumentType"/>, and nothing here reaches it.</para>
///
/// <para>A payment request may pay several invoices at once — the same supplier billing two plants,
/// a monthly service split per site. Each document keeps its own identity, OCR reading,
/// classification decision, dates, values and items, because each derives its own post-payment
/// obligations. Aggregating them into one request-level type would make a request that pays a
/// proforma AND an invoice inexpressible.</para>
///
/// <para>This is <b>not</b> an <see cref="OperationInvoice"/>. A source document initiates payment;
/// an operation invoice arrives later to discharge an obligation the source document created. They
/// have different lifecycles, permissions and audits, and merging them is how "final invoice"
/// became ambiguous in the first place.</para>
/// </summary>
public class PaymentSourceDocument
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public Guid RequestId { get; set; }
    public Request? Request { get; set; }

    /// <summary>The file itself. Unique — one attachment is one source document.</summary>
    public Guid AttachmentId { get; set; }
    public RequestAttachment? Attachment { get; set; }

    // ── Identity of the commercial counterpart ──
    public int? SupplierId { get; set; }
    public Supplier? Supplier { get; set; }
    public string? SupplierNameSnapshot { get; set; }
    public string? SupplierTaxIdSnapshot { get; set; }

    /// <summary>
    /// The plant this document is for. Part of the PO-group key, which is why two invoices from one
    /// supplier for two plants produce two groups.
    /// </summary>
    public int? PlantId { get; set; }
    public Plant? Plant { get; set; }

    // ── Document identity (Release 2 taxonomy) ──
    /// <summary>What the supplier issued — see RequestConstants.SourceDocumentTypes.</summary>
    public string? SourceDocumentType { get; set; }
    public string? DocumentNumber { get; set; }
    public string? DocumentSeries { get; set; }
    public DateTime? DocumentDate { get; set; }
    public DateTime? DueDate { get; set; }
    public string? Currency { get; set; }

    public decimal? NetAmount { get; set; }
    public decimal? TaxAmount { get; set; }
    public decimal? GrossAmount { get; set; }

    // ── How the classification was reached (Release 2 evidence model, per document) ──
    public string? OcrSuggestion { get; set; }
    public decimal? OcrConfidence { get; set; }
    public string? OcrEvidenceJson { get; set; }
    public string? OcrConflictingEvidenceJson { get; set; }
    public string? OcrTitleFound { get; set; }

    /// <summary>USER_SELECTED, OCR_CONFIRMED or FINANCE_REVIEW.</summary>
    public string? ClassificationSource { get; set; }
    /// <summary>OCR or FALLBACK — how the suggestion was reached.</summary>
    public string? ClassificationSuggestionSource { get; set; }
    /// <summary>The user was shown a contradiction with the reading and proceeded anyway.</summary>
    public bool ClassificationConflictAcknowledged { get; set; }
    /// <summary>Mandatory written reason when a high-risk contradiction was overridden.</summary>
    public string? ClassificationJustification { get; set; }

    /// <summary>Finance reviewed this document's classification, where the type requires it.</summary>
    public bool ClassificationReviewedByFinance { get; set; }
    public Guid? ClassificationReviewedByUserId { get; set; }
    public DateTime? ClassificationReviewedAtUtc { get; set; }

    // ── Position and lifecycle ──
    /// <summary>1..N within the request — stable "Documento 2" labelling that survives a removal.</summary>
    public int SequenceNumber { get; set; }

    /// <summary>
    /// Removed after submission. Never hard-deleted at that point: its classification decision and
    /// any justification were already audited, and an audit must survive the object it describes.
    /// Before submission the row is deleted outright — nothing downstream exists yet.
    /// </summary>
    public bool IsVoided { get; set; }
    public DateTime? VoidedAtUtc { get; set; }
    public Guid? VoidedByUserId { get; set; }
    public string? VoidReason { get; set; }

    public DateTime CreatedAtUtc { get; set; }
    public Guid CreatedByUserId { get; set; }
    public DateTime? UpdatedAtUtc { get; set; }
    public Guid? UpdatedByUserId { get; set; }

    public byte[] RowVersion { get; set; } = Array.Empty<byte>();

    /// <summary>
    /// The items this document pays for. Their <c>RequestPoGroupId</c> is what distributes the
    /// document across PO groups — no separate allocation table exists, because a stored allocation
    /// amount could disagree with the sum of the lines it claims to represent.
    /// </summary>
    public ICollection<RequestLineItem> LineItems { get; set; } = new List<RequestLineItem>();
}
